docs/annotated_mnist.md

```diff
@@ -1,157 +1,317 @@
 
-```py
-class CNN(flax.nn.Module):
-  def apply(self, x):
-    x = flax.nn.Conv(x, features=32, kernel_size=(3, 3))
-    x = flax.nn.relu(x)
-    x = flax.nn.avg_pool(x, window_shape=(2, 2), strides=(2, 2))
-    x = flax.nn.Conv(x, features=64, kernel_size=(3, 3))
-    x = flax.nn.relu(x)
-    x = flax.nn.avg_pool(x, window_shape=(2, 2), strides=(2, 2))
-    x = x.reshape((x.shape[0], -1))
-    x = flax.nn.Dense(x, features=256)
-    x = flax.nn.relu(x)
-    x = flax.nn.Dense(x, features=10)
-    x = flax.nn.log_softmax(x)
-    return x
+
+```python
+def create_optimizer(params, learning_rate, beta):
+  optimizer_def = optim.Momentum(learning_rate=learning_rate, beta=beta)
+  optimizer = optimizer_def.create(params)
+  return optimizer
 ```
 
-`jax.vmap` allows us to define the `cross_entropy_loss`
-function as if it acts on a single sample. `jax.vmap`
-automatically vectorizes code efficiently to run on entire
-batches.
+3. Create a function for parameter initialization:
 
-```py
-@jax.vmap
-def cross_entropy_loss(logits, label):
-  return -logits[label]
+  - Set the initial shape of the kernel (note that JAX and Flax are [row-based](https://flax.readthedocs.io/en/latest/notebooks/flax_basics.html#Model-parameters-&-initialization)); and
+  - Initialize the module parameters of your network (`CNN`) with the [`flax.linen.Module.init`](https://flax.readthedocs.io/en/latest/flax.linen.html#flax.linen.Module.init) method using the PRNGKey and parameters (note that the parameters aren't stored with the models).
+
+
+```python
+def get_initial_params(key):
+  init_shape = jnp.ones((1, 28, 28, 1), jnp.float32)
+  initial_params = CNN().init(key, init_shape)['params']
+  return initial_params
 ```
 
-Compute loss and accuracy. We use `jnp` (`jax.numpy`) which can run on
-device (GPU or TPU).
+4. For loss and accuracy metrics, create a separate function:
 
-```py
+
+```python
 def compute_metrics(logits, labels):
-  loss = jnp.mean(cross_entropy_loss(logits, labels))
+  loss = cross_entropy_loss(logits, labels)
   accuracy = jnp.mean(jnp.argmax(logits, -1) == labels)
-  return {'loss': loss, 'accuracy': accuracy}
+  metrics = {
+      'loss': loss,
+      'accuracy': accuracy
+  }
+  return metrics
 ```
 
-`jax.jit` traces the `train_step` function and compiles into fused device
-operations that run on GPU or TPU.
+## The dataset
+
+Define a function that:
+  - Uses TFDS to load and prepare the MNIST dataset; and
+  - Converts the samples to floating-point numbers.
+
+
+```python
+def get_datasets():
+  ds_builder = tfds.builder('mnist')
+  ds_builder.download_and_prepare()
+  # Split into training/test sets
+  train_ds = tfds.as_numpy(ds_builder.as_dataset(split='train', batch_size=-1))
+  test_ds = tfds.as_numpy(ds_builder.as_dataset(split='test', batch_size=-1))
+  # Convert to floating-points
+  train_ds['image'] = jnp.float32(train_ds['image']) / 255.0
```

Comment:
It's common to use this better conversion logic: `(int_val + 0.5) / 256.0`

Reply:
Cool. Should we update the MNIST example in `linen_examples`? The code comes directly from there (that's how I started learning Linen 🙂).

https://github.com/google/flax/blob/9fdf59f05702c746ea9a6be50cd3a81f38e2fa0b/linen_examples/mnist/train.py#L138

```py
def get_datasets():
  """Load MNIST train and test datasets into memory."""
  ds_builder = tfds.builder('mnist')
  ds_builder.download_and_prepare()
  train_ds = tfds.as_numpy(ds_builder.as_dataset(split='train', batch_size=-1))
  test_ds = tfds.as_numpy(ds_builder.as_dataset(split='test', batch_size=-1))
  train_ds['image'] = jnp.float32(train_ds['image']) / 255.
  test_ds['image'] = jnp.float32(test_ds['image']) / 255.
  return train_ds, test_ds
```